File: ruby/correct.rb
require( 'json' )
require_relative( './helpers' )

module OTX
  class StatementCorrector
    include OTX::Helpers
    
    # Build a GraphQL mutation for updating (correcting) statements.
    # @param corrections [Array] An array changes to make.  Each set of 
    #   changes should be an uploader_id and a set of field/value pairs.
    #   See the format in ../data/f1099nec-corrections.json.
    # @return [String] The full GraphQL update mutation.
    def build_mutation( corrections: )
      data = corrections.map{ |s| to_gql( object: s ) }.join( "\n" )
      <<-END_MUTATION
        mutation {
          updateF1099necStatements(
            statements: [ #{data} ]
          ) {
            statements {
              recordNumber
              statement {
                otxId
                uploaderId
                nonemployeeComp
                recipientCity
                recipientZipCode
              }
              messages
            }
            errors
          }
        }      
      END_MUTATION
    end
    def correct
      credential = get_credential
      # See ../data/f1099nec-corrections.json for correction format.
      corrections = get_data( filename: '../data/f1099nec-corrections.json' )
      mutation = build_mutation(corrections: corrections)
      response = post_gql( credential: credential,
                           payload: mutation )
      # response is a hash that you can manipulate to suit your needs.
      # Here we're just going to print it (as a JSON object to make it
      # easier to read).
      STDOUT.puts( JSON.pretty_generate( response ) )
      # Credentials are valid for several days after they've been issued.
      # If you plan to do more work, you can use the same credential.  When done 
      # working, you should delete the credential to provide additional security.
      if( delete_credential( credential: credential ) )
        STDOUT.puts( "\nSuccessfully deleted credential." )
      end
    end
  end
end

corrector = OTX::StatementCorrector.new
corrector.correct

File: python/correct.py
import json
from helper import Helper

class StatementCorrector:
    def __init__(self):
        self.helper = Helper()
  
    def build_mutation(self, corrections):
        """Build a GraphQL mutation for updating (correcting) statements.

        :param list corrections: A list of changes to make.  Each set of 
          changes should be an uploader_id and a set of field/value pairs.
          See the format in ../data/f1099nec-corrections.json.
        :return: The full GraphQL update mutation
        :rtype: str
        """

        data = "\n".join(map(self.helper.to_gql, corrections))
        str = """
          mutation {
            updateF1099necStatements(
              statements: [ %s ]
            ) {
              statements {
                recordNumber
                statement {
                  otxId
                  uploaderId
                  nonemployeeComp
                  recipientCity
                  recipientZipCode
                }
                messages
              }
              errors
            }
          }      
        """ % (data)
        return(str)

    def correct(self):
        """Correct statements using data in ../data/f1099nec-corrections.json"""
        credential = self.helper.get_credential()
        # See ../data/f1099nec-corrections.json for correction format.
        corrections = self.helper.get_data('../data/f1099nec-corrections.json')
        mutation = self.build_mutation(corrections)
        response = self.helper.post_gql( credential, mutation )
        # response is a dict that you can manipulate to suit your needs.
        # Here we're just going to print it (as a JSON object to make it
        # easier to read).
        print(json.dumps(response, indent = 2))
        # Credentials are valid for several days after they've been issued.
        # If you plan to do more work, you can use the same credential.  When done 
        # working, you should delete the credential to provide additional security.
        if(self.helper.delete_credential(credential)):
            print( "\nSuccessfully deleted credential." )

corrector = StatementCorrector()
corrector.correct()
